### PR TITLE
Printing producer pip description when we have multiple consumers

### PR DESCRIPTION
**Pull request by rijulluman-msft**
[AB#1586413](https://dev.azure.com/mseng/708e929f-6bd5-415a-8daf-25b1dac08dd8/_workitems/edit/1586413)

### CODE REVIEW

File: Public/Src/Engine/Scheduler/Artifacts/FileContentManager.cs [hunk 1]
```diff
@@ -2818,7 +2818,7 @@ private string GetAssociatedPipDescription(FileOrDirectoryArtifact declaredArtif
                 if (declaredArtifact.FileArtifact.IsSourceFile)
                 {
                     consumerDescription = m_host.GetConsumerDescription(declaredArtifact);
```
Comment (ljcollins25) [inline Public/Src/Engine/Scheduler/Artifacts/FileContentManager.cs]: >GetConsumerDescription [](start = 49, length = 22)

This is the only place where this method is called. Maybe change its behavior to return null if there is more than one consumer (and update comment on IFileContentManagerHost.GetConsumerDescription to represent new behavior of the API.

File: Public/Src/Engine/Scheduler/Scheduler.cs [hunk 2]
```diff
@@ -5476,6 +5476,29 @@ public string GetConsumerDescription(in FileOrDirectoryArtifact artifact)
             return null;
         }
 
+        /// <summary>
+        /// Returns true when associated with a FileOrDirectory artifact has more than 1 consumers.
+        /// </summary>
+        /// <param name="artifact">The artifact for which to find consumer count > 1.</param>
+        /// <returns>The true when consumer count = 0 or 1 or false if consumer count > 1.</returns>
+        [SuppressMessage("Microsoft.Design", "CA1033:InterfaceMethodsShouldBeCallableByChildTypes")]
+        public bool HasMultipleConsumers(in FileOrDirectoryArtifact artifact)
```
Comment (ljcollins25) [inline Public/Src/Engine/Scheduler/Scheduler.cs]: >HasMultipleConsumers [](start = 20, length = 20)

Per my other suggestion, this method is no longer needed, but for future reference you can get the number of outgoing/incoming edges by calling DataflowGraph.GetOutgoingEdgesCount(nodeId)

File: Public/Src/Engine/Scheduler/Scheduler.cs [hunk 3]
```diff
@@ -5458,22 +5458,26 @@ string IFileContentManagerHost.GetProducerDescription(in FileOrDirectoryArtifact
...
         /// <returns>The first consumer description or null if there is no consumer.</returns>
         [SuppressMessage("Microsoft.Design", "CA1033:InterfaceMethodsShouldBeCallableByChildTypes")]
         public string GetConsumerDescription(in FileOrDirectoryArtifact artifact)
         {
             var producerId = PipGraph.GetProducer(artifact);
+            string singleConsumer = null;
             foreach (var consumerEdge in DataflowGraph.GetOutgoingEdges(producerId.ToNodeId()))
             {
+                if (singleConsumer != null) {
+                    return null;    // More than 1 consumers
```
Comment (olkononenko) [inline Public/Src/Engine/Scheduler/Scheduler.cs]: >// More than 1 consumers [](start = 36, length = 24)

I don't think this is a right approach. Now all the callers need to handle this 'null'. 
This method is called in `GetAssociatedPipDescription` - multiple consumers -> returning `null` -> the caller will have to search for description of a producer... How does it align with the error logging issue this PR claims to be addressing?

File: Public/Src/Engine/Scheduler/Scheduler.cs [hunk 4]
```diff
@@ -5458,22 +5458,26 @@ string IFileContentManagerHost.GetProducerDescription(in FileOrDirectoryArtifact
...
         /// <param name="artifact">The artifact for which to get the first consumer description.</param>
         /// <returns>The first consumer description or null if there is no consumer.</returns>
         [SuppressMessage("Microsoft.Design", "CA1033:InterfaceMethodsShouldBeCallableByChildTypes")]
         public string GetConsumerDescription(in FileOrDirectoryArtifact artifact)
         {
             var producerId = PipGraph.GetProducer(artifact);
+            string singleConsumer = null;
             foreach (var consumerEdge in DataflowGraph.GetOutgoingEdges(producerId.ToNodeId()))
             {
+                if (singleConsumer != null) {
```
Comment (almiliMSFT) [inline Public/Src/Engine/Scheduler/Scheduler.cs]: nit:
```suggestion
                if (singleConsumer != null)
                {
```

File: Public/Src/Engine/Scheduler/Scheduler.cs (hunk 3 above)
Comment (almiliMSFT) [reply Public/Src/Engine/Scheduler/Scheduler.cs]: The callers had to handle `null` before too

File: Public/Src/Engine/Scheduler/Scheduler.cs [hunk 5]
```diff
@@ -5458,22 +5458,26 @@ string IFileContentManagerHost.GetProducerDescription(in FileOrDirectoryArtifact
...
+                }
                 Pip consumer = PipGraph.GetPipFromPipId(consumerEdge.OtherNode.ToPipId());
-                return consumer.GetDescription(Context);
+                singleConsumer = consumer.GetDescription(Context);
             }
 
-            // No consumer
-            return null;
+            // Single consumer description or null for 0 consumers
+            return singleConsumer;
```
Comment (almiliMSFT) [inline Public/Src/Engine/Scheduler/Scheduler.cs]: I don't know if this is the right change, but if it is, can you please write it in a less convoluted way, e.g.,:
```csharp
    var producerId = PipGraph.GetProducer(artifact);
    var numConsumers = DataflowGraph.GetOutgoingEdgesCount(producerId.ToNodeId());
    if (numConsumers == 1)
    {
        var consumerPipId = DataflowGraph.GetOutgoingEdges(producerId.ToNodeId()).First().OtherNode.ToPipId();
        var consumer = PipGraph.GetPipFromPipId(consumerPipId);
        return consumer.GetDescription(Context);
    }
    else
    {
        return null;
    }
```
No `for` loop is needed if we only care about the first consumer and/or whether there exist more than 1 consumer.

**Pull request closed without merging**